Announcements: Cleanup and remove hardcoded "xbmc" string in a lot of code

a1rwulf opened the pull request:
## Description
In all Announcements "xbmc" was hardcoded as the sender.
Also the interface was using `const char*`.
This PR moves the interface to use `std::string` and introduces a constant
in `CAnnouncementManager`.
Additionally, as so much code just uses xbmc as a sender we offer new convenience methods
that imply the sender.

## Motivation and Context
Cleanup of const char* usage and getting rid of hardcoded "xbmc" strings.

## How Has This Been Tested?
Did a short test on linux.
Player notifications unchanged.
Actually no user facing functionality should change at all.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

a1rwulf committed:
Cosmetics: ran git-clang-format

ksooo reviewed (changes requested):
Sorry, this is a nogo, since the string "xbmc" is part of the public API and does not denote the application name here! Many addons rely on "xbmc" as sender name and would break immediately and even worse would have to be adapted for every "app" they will ever get notifications from.

a1rwulf committed:
Annoncements: Make interface use std::string

a1rwulf commented:
Removed the breaking change label, because after reworking the PR and thus reverting the use of `AppName` it's not a breaking change anymore.

ksooo reviewed (approved):
Very nice. Much appreciated.

ksooo commented:
Before you merge, could you please check that all of the old overloads of Announce (taking a sendet parameter) are still used? If not, please remove the unused code.

ksooo commented:
And another thing that just came to my mind: if ANNOUNCEMENT_SENDER is not needed outside the cop file, there is no need to expose it via the header. Better choice is a static var wrapped in an unnamed namespace in the cpp. And in this case it even can be constexpr which gives the computer further optimization possibilities.

a1rwulf commented:
> And another thing that just came to my mind: if ANNOUNCEMENT_SENDER is not needed outside the cop file, there is no need to expose it via the header. Better choice is a static var wrapped in an unnamed namespace in the cpp. And in this case it even can be constexpr which gives the compiler further optimization possibilities.

It's used on some places outside like here:
https://github.com/xbmc/xbmc/pull/16760/commits/0e43b040aff65e14445ce51792896851bdfc91c9#diff-e338afa8dde3f983a08158e658bd52d0R161

ksooo commented:
> It's used on some places outside like here:

 Ah. Ofc. Sorry for the noise.

arnova commented on xbmc/windows/GUIWindowHome.cpp:
```diff
@@ -74,11 +77,10 @@ void CGUIWindowHome::Announce(ANNOUNCEMENT::AnnouncementFlag flag, const char *s
     return;
 
-  bool onUpdate = strcmp(message, "OnUpdate") == 0;
+  bool onUpdate = message == "OnUpdate";
```
Please use brackets like (message == "OnUpdate"), it improves readability. And personally I would mark onUpdate const.

a1rwulf replied:
We have those hardcoded strings all over and I don't want to "pollute" this PR with changing another 100 of string literal replaces - I'll probably do that in a separate PR.
I prefer to not have the brackets because I think we use that style on a lot of other places as well.

a1rwulf commented:
@ksooo Removed one method that's now not used anymore and made sure the new convenience methods call directly in the actual implementation instead of going through the second interface method that has the sender as param.

a1rwulf commented:
jenkins build this please

a1rwulf commented:
jenkins build this please

wsnipex commented:
This has potential to break remote apps and home automation stuff, so might warrant a heads up. Did you test this with Kore?

Rechi commented on xbmc/network/upnp/UPnPRenderer.cpp:
```diff
@@ -227,57 +228,67 @@ CUPnPRenderer::ProcessHttpGetRequest(NPT_HttpRequest&              request,
+      rct->SetStateVariable("VolumeDb", buffer.c_str());
 
-        rct->SetStateVariable("Mute", data["muted"].asBoolean() ? "1" : "0");
+      rct->SetStateVariable("Mute", data["muted"].asBoolean() ? "1" : "0");
```
indentation is wrong

a1rwulf commented:
> This has potential to break remote apps and home automation stuff, so might warrant a heads up. Did you test this with Kore?

@wsnipex I just did a test with kore.
Controlling still works and also playback notifications seem to be fine, so I tend to think I didn't break anything.

a1rwulf committed:
Announcements: Get rid of hardcoded "xbmc" in code

Because 90% of the cases hardcode "xbmc" as a sender
we now offer a method in the interface that will
imply the sender "xbmc".
Furthermore we now have a constant string in
AnnouncementManager that can be used.

a1rwulf committed:
Cosmetics: ran git-clang-format

a1rwulf commented:
jenkins build this please

garbear reviewed (approved):
I prefer the extra brackets. I super prefer not touching existing code. LGTM

a1rwulf commented:
jenkins build this please
@Rechi any idea why jenkins doesn't build this PR?

a1rwulf commented:
jenkins build this please

a1rwulf committed:
Announcements: Remove unused method and remove one layer of indirection

phunkyfish commented:
@a1rwulf do you want to rebase and hopefully jenkins will build it now?

a1rwulf commented:
It will still fail on iOS, seems I also need to change some obj-c files - haven't had time to look into it yet.

fuzzard commented on xbmc/interfaces/AnnouncementManager.cpp:
```diff
@@ -25,6 +25,8 @@
 
 using namespace ANNOUNCEMENT;
 
+const std::string CAnnouncementManager::ANNOUNCEMENT_SENDER = "xbmc";
```
So removing the xbmc from all the announcements is great, should this be taken a step further, and rather than hardcode this as "xbmc" should we use 

https://github.com/xbmc/xbmc/blob/94184731d2bb21f98d5964a268e18cc2a04ffb9d/xbmc/CompileInfo.cpp.in#L43

garbear replied:
Without looking at the code, I can say usually not, because the cast from char* to std::string requires an alloc anywhere the string is used, whereas instantiating as std::string upfront gives you data that can be used later without allocations.

fuzzard replied:
couldnt you just static const std::string ... = CCompileInfo::GetAppName()?

garbear replied:
is it specified as static: https://github.com/xbmc/xbmc/blob/1d40c5a3e702e0bcb919d1a9ee6dda360eb3661d/xbmc/interfaces/AnnouncementManager.h#L61

for linking, it has to be defined in a compilation unit with the rest of the class

a1rwulf replied:
> couldnt you just static const std::string ... = CCompileInfo::GetAppName()?
> 
> The rest of this PR essentially drops the sender being passed, and this is then just assumed in the announcementmanager to be this const

See comment below.
There seem to be a bunch of addons and/or json clients that check on their side if the sender is "xbmc".
Changing from "xbmc" to GetAppName() would break all of them, guess it's not worth the trouble.

fuzzard commented on xbmc/interfaces/AnnouncementManager.h:
```diff
@@ -34,17 +34,43 @@ namespace ANNOUNCEMENT
+
+    // The sender is not related to the application name.
+    // Also it's part of Kodi's API - changing it will break
+    // a big number of python addons and third party json consumers.
```
considering what else we are doing around python api changes, should we take the opportunity to change this?

a1rwulf replied:
I think it's not worth it, just to have "kodi" instead of "xbmc" in the json notifications.

fuzzard replied:
Thats fine, its not about "kodi" or "xbmc", its more the trademark stuff for rebranding thats been worked on. Im not across all that, so maybe a question for @Darren-Hill or some of the board members who have been dealing with that stuff.
If its going to cause a breaking change, and for trademark purposes it needs to be changed, probably best to do with our already existing breaking changes. if its not deemed an issue, sorry for the noise

a1rwulf replied:
Afaik it's not needed for trademark purposes, otherwise it should have been part of the kodi rebrand already - it's not really user facing, just some internal string that is sent to a socket, could be "asdf" as well - I bet most json clients don't care about it at all, at least I don't really understand why it's there at all.
Normally if I as a client connect to a ip/socket pair, I already know which service I connect to, so why would I need a "sender" identifier?
Only one kodi/xbmc/rebranded-fork can bind that socket anyway.

Darren-Hill replied:
A similar question has actually come up as part of the trademark/rebranding work, as for binary addons if you rebrand and change the app name the binary addons do still get installed in a Kodi directory rather than using the new rebrand name. @ronie did some work on this and that was the only issue that got identified from it (via changing APP_NAME in version.txt). 

See his comment on 16th Dec in [this issue](https://github.com/xbmc/board/issues/118) for more details of what he found.

As noted it is questionable whether usage of xbmc/Kodi within the code itself it subject to trademark or not (or indeed if such uses can even be trademarked), but the above is a known blocking point to making rebranding easier for those users who want to go that way.

a1rwulf replied:
Yeah, while above is probably a bigger issue for rebranders (so is that you need to change include paths in each and every binary addon you want to use), the question in discussion is 
neglectable from a rebranders POV - we could simply add it to the rebrand guide that you can change this new constant if you don't want to have the term "xbmc" in your rebrands json notifications -> probably remotes like KORE etc. will not work with your rebrand then.

The question in matter is:
Was it an oversight, that we as team kodi, need to rebrand this string, or is it save to leave it as is.

fuzzard commented:
ios looks clean regarding announcement stuff. Only reference i found in darwin platform stuff (that you havent covered already) was in tvos. Sorry for introducing it, haha

https://github.com/xbmc/xbmc/blob/e583c65abd3fb512a449f80e723f2b900292c350/xbmc/platform/darwin/tvos/XBMCController.mm#L317

a1rwulf commented:
Finally found some time and motivation to rebase this thing :)

a1rwulf commented:
Jenkins is finally happy.
@ksooo Do you prefer keeping the code formatting in separate commits or shall I squash them as well?

ksooo commented:
> Do you prefer keeping the code formatting in separate commits or shall I squash them as well?

 No hard feelings here. Your decision.
